### PR TITLE
[multi asic] multi asic changes for traffic_shift test

### PR DESCRIPTION
**Pull request by judyjoseph**
### Description of PR
This PR addresses the following changes 

1. For multi-asic the console output changes to the following format, below is the case where there are **4 front-end ASIC's**

```
admin@str--acs-1:~$ TSA
BGP0 : System Mode: Normal -> Maintenance
BGP1 : System Mode: Normal -> Maintenance
BGP2 : System Mode: Normal -> Maintenance
BGP3 : System Mode: Normal -> Maintenance
admin@str--acs-1:~$ TSC
BGP0 : System Mode: Maintenance
BGP1 : System Mode: Maintenance
BGP2 : System Mode: Maintenance
BGP3 : System Mode: Maintenance
admin@str--acs-1:~$ TSB
BGP0 : System Mode: Maintenance -> Normal
BGP1 : System Mode: Maintenance -> Normal
BGP2 : System Mode: Maintenance -> Normal
BGP3 : System Mode: Maintenance -> Normal
admin@str--acs-1:~$ TSC
BGP0 : System Mode: Normal
BGP1 : System Mode: Normal
BGP2 : System Mode: Normal
BGP3 : System Mode: Normal
```
2. Also added additional testcase **test_TSA_B_C_with_no_neighbors** to handle the new console message in case there are **no external BGP neighbors** for that BGP instance. This is more predominantly seen in multi-asic where one ASIC may have no interfaces up and hence no external BGP neighbors Ref: https://github.com/Azure/sonic-buildimage/pull/7171


```
admin@STG01-0101-0102-01T1:~$ TSA
BGP0 : System Mode: Normal -> Maintenance
BGP1 : System Mode: Normal -> Maintenance
BGP2 : System Mode: No external neighbors
BGP3 : System Mode: Normal -> Maintenance 
```

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Approach
#### What is the motivation for this PR?

#### How did you do it?

#### How did you verify/test it?
The tests was run against single-asic/ multi-asic platform with result PASS

```
Single ASIC DUT

jujoseph@63d6771d2f5f:/var/mgmt/Networking-acs-sonic-mgmt/tests$ ./run_tests.sh -d str-dx010-acs-1 -n vms3-t1-dx010-1 -i ../ansible/veos,../ansible/str -f ../ansible/testbed.csv -c  bgp/test_traffic_shift.py -u -e "--skip_sanity"
=== Running tests in groups ===
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
=================================================================================================================== test session starts ===================================================================================================================
platform linux2 -- Python 2.7.17, pytest-4.6.5, py-1.10.0, pluggy-0.13.1
ansible: 2.8.12
rootdir: /var/mgmt/Networking-acs-sonic-mgmt/tests, inifile: pytest.ini
plugins: forked-1.3.0, metadata-1.11.0, xdist-1.28.0, html-1.22.1, repeat-0.9.1, ansible-2.2.2
collected 3 items                                                                                                                                                                                                                                         

bgp/test_traffic_shift.py::test_TSA PASSED                                                                                                                                                                                                          [ 33%]
bgp/test_traffic_shift.py::test_TSB PASSED                                                                                                                                                                                                          [ 66%]
bgp/test_traffic_shift.py::test_TSA_B_C_with_no_neighbors PASSED                                                                                                                                                                                    [100%]

---------------------------------------------------------------------------------------- generated xml file: /var/mgmt/Networking-acs-sonic-mgmt/tests/logs/tr.xml ----------------------------------------------------------------------------------------
=============================================================================================================== 3 passed in 907.57 seconds ================================================================================================================

multi-asic DUT 

jujoseph@63d6771d2f5f:/var/mgmt/Networking-acs-sonic-mgmt/tests$ ./run_tests.sh -d str2--acs-1 -n vms17-t1--acs-1 -i ../ansible/veos,../ansible/str2 -f ../ansible/testbed.csv -c  bgp/test_traffic_shift.py -u -e "--skip_sanity"
=== Running tests in groups ===
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
=================================================================================================================== test session starts ===================================================================================================================
platform linux2 -- Python 2.7.17, pytest-4.6.5, py-1.10.0, pluggy-0.13.1
ansible: 2.8.12
rootdir: /var/mgmt/Networking-acs-sonic-mgmt/tests, inifile: pytest.ini
plugins: forked-1.3.0, metadata-1.11.0, xdist-1.28.0, html-1.22.1, repeat-0.9.1, ansible-2.2.2
collected 3 items                                                                                                                                                                                                                                         

bgp/test_traffic_shift.py::test_TSA PASSED                                                                                                                                                                                                          [ 33%]
bgp/test_traffic_shift.py::test_TSB PASSED                                                                                                                                                                                                          [ 66%]
bgp/test_traffic_shift.py::test_TSA_B_C_with_no_neighbors PASSED                                                                                                                                                                                    [100%]

---------------------------------------------------------------------------------------- generated xml file: /var/mgmt/Networking-acs-sonic-mgmt/tests/logs/tr.xml ----------------------------------------------------------------------------------------
=============================================================================================================== 3 passed in 814.77 seconds ================================================================================================================
```



#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation 
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

### CODE REVIEW

**Comment by smaheshm** [comment]
Why not use "sonic_asic" object to do the work instead of checking for multi ASIC host, and evaluating the namespace every time. Add a BGP command API in the "sonic_asic.py" that takes of appending the "vtysh -n <asic>". 

you can get sonic_asic object by "duhost.asic_instance(asic_index)"

File: tests/bgp/bgp_helpers.py [hunk 1]
```diff
@@ -99,13 +99,15 @@ def parse_rib(host, ip_ver):
...
-    route_data = json.loads(host.shell(cmd, verbose=False)['stdout'])
-    for ip, nexthops in route_data['routes'].iteritems():
-        aspath = set()
-        for nexthop in nexthops:
-            aspath.add(nexthop['path'])
-        routes[ip] = aspath
+    for asic_index in host.get_frontend_asic_ids():
+        namespace = '-n {}'.format(asic_index) if host.is_multi_asic else ''
+        cmd = "vtysh %s -c \"show bgp ipv%d json\"" % (namespace, ip_ver)
+        route_data = json.loads(host.shell(cmd, verbose=False)['stdout'])
```
Comment (abdosi) [inline tests/bgp/bgp_helpers.py:105]: why we need to walk for all asics ? show bgp gets data from all asic. will that not work.

**Comment by judyjoseph** [comment]
> Why not use "sonic_asic" object to do the work instead of checking for multi ASIC host, and evaluating the namespace every time. Add a BGP command API in the "sonic_asic.py" that takes of appending the "vtysh -n ".
> 
> you can get sonic_asic object by "duhost.asic_instance(asic_index)"

Updated to use the API get_vtysh_cmd_for_namespace(). I felt adding the others like in the function verify_traffic_shift_per_asic()/verify_traffic_shift()/get_traffic_shift_state() in "sonic_asic.py" might shift many BGP specific string definitions there in the common  code ?

File: tests/bgp/bgp_helpers.py (hunk 1 above)
Comment (judyjoseph) [reply tests/bgp/bgp_helpers.py:111]: I did try that earlier .. but have to use this way as the "show ip route" even with verbose option don't dump the aspath info. The aspath info is used later in the verify_* API's in test_traffic_shift.py script.

File: tests/bgp/bgp_helpers.py [hunk 2]
```diff
@@ -32,15 +34,17 @@ def apply_bgp_config(duthost, template_name):
...
 
     Args:
         duthost: DUT host object
     """
-    duthost.shell('systemctl restart bgp')
-    time.sleep(60)
+    duthost.asic_instance(asic_index).reset_service("bgp")
+    duthost.asic_instance(asic_index).restart_service("bgp")
+    docker_name = duthost.asic_instance(asic_index).get_docker_name("bgp")
+    pytest_assert(wait_until(100, 10, duthost.is_service_fully_started, docker_name), "BGP not started.")
```
Comment (smaheshm) [inline tests/bgp/bgp_helpers.py:47]: 👍 replacing sleep with 'wait_until'

File: tests/bgp/conftest.py [hunk 3]
```diff
@@ -220,29 +220,41 @@ def _setup_interfaces_t1(mg_facts, peer_count):
             _subnets = ipaddress.ip_network(u"10.0.0.0/24").subnets(new_prefix=subnet_prefixlen)
             subnets = (_ for _ in _subnets if _ not in used_subnets)
 
+            loopback_ip = None
+            for intf in mg_facts["minigraph_lo_interfaces"]:
+                if netaddr.IPAddress(intf["addr"]).version == 4:
+                    loopback_ip = intf["addr"]
+                    break
+                if not loopback_ip:
```
Comment (smaheshm) [inline tests/bgp/conftest.py:228]: shouldn't this be outside of for loop?

File: tests/bgp/conftest.py [hunk 4]
```diff
@@ -220,29 +220,41 @@ def _setup_interfaces_t1(mg_facts, peer_count):
...
                 else:
                     conn["neighbor_intf"] = "eth%s" % mg_facts["minigraph_port_indices"][intf]
                 connections.append(conn)
 
             for conn in connections:
                 # bind the ip to the interface and notify bgpcfgd
-                duthost.shell("config interface ip add %s %s" % (conn["local_intf"], conn["local_addr"]))
+                namespace = '-n {}'.format(conn["namespace"]) if conn["namespace"] else ''
+                duthost.shell("config interface %s ip add %s %s" % (namespace, conn["local_intf"], conn["local_addr"]))
                 ptfhost.shell("ifconfig %s %s" % (conn["neighbor_intf"], conn["neighbor_addr"]))
```
Comment (smaheshm) [inline tests/bgp/conftest.py:250]: doesn't this need namespace arg?

Comment (smaheshm) [reply tests/bgp/conftest.py:250]: never mind, didn't realize it's the ptfhost.

File: tests/bgp/test_traffic_shift.py [hunk 5]
```diff
@@ -173,3 +190,42 @@ def test_TSB(duthost, ptfhost, nbrhosts, bgpmon_setup_teardown):
...
+
+        # Issue TSA on DUT
+        output = duthost.shell("TSA")['stdout_lines']
+
+        # Set the DUT in maintenance state
+        # Verify ASIC0 has no neighbors message.
+        pytest_assert(verify_traffic_shift_per_asic(duthost, output, TS_NO_NEIGHBORS, asic_index), "ASIC is not having no neighbors")
+
+        # Recover to Normal state
+        # Verify ASIC0 has no neighbors message.
```
Comment (smaheshm) [inline tests/bgp/test_traffic_shift.py:213]: remove the comment.

File: tests/bgp/test_traffic_shift.py [hunk 6]
```diff
@@ -173,3 +190,42 @@ def test_TSB(duthost, ptfhost, nbrhosts, bgpmon_setup_teardown):
...
+        duthost.shell("TSB")['stdout_lines']
+
+        # Verify DUT is in Normal state, and ASIC0 has no neighbors message.
+        pytest_assert(verify_traffic_shift_per_asic(duthost, output, TS_NO_NEIGHBORS, asic_index), "ASIC is not having no neighbors")
+
+        # Check the traffic state
+        # Verify ASIC0 has no neighbors message.
```
Comment (smaheshm) [inline tests/bgp/test_traffic_shift.py:220]: remove the comment.

File: tests/bgp/conftest.py (hunk 3 above)
Comment (judyjoseph) [reply tests/bgp/conftest.py:228]: Sure will update .. this also works as there is only one loopback interface.

File: tests/bgp/test_traffic_shift.py [hunk 7]
```diff
@@ -1,35 +1,52 @@
...
 TS_INCONSISTENT = "System Mode: Not consistent"
+TS_NO_NEIGHBORS = "System Mode: No external neighbors"
 
 @pytest.fixture
 def traffic_shift_community(duthost):
     community = duthost.shell('sonic-cfggen -y /etc/sonic/constants.yml -v constants.bgp.traffic_shift_community')['stdout']
     return community
 
+def verify_traffic_shift_per_asic(host, outputs, match_result, asic_index):
+    prefix = "BGP{} : ".format(asic_index) if asic_index != DEFAULT_ASIC_ID else ''
```
Comment (abdosi) [inline tests/bgp/test_traffic_shift.py:27]: TSA/B/C being system level feature why we want per asic validation ?

File: tests/bgp/test_traffic_shift.py [hunk 8]
```diff
@@ -1,35 +1,52 @@
...
-from bgp_helpers import parse_rib, verify_all_routes_announce_to_bgpmon 
+from bgp_helpers import parse_rib, verify_all_routes_announce_to_bgpmon,remove_bgp_neighbors,restore_bgp_neighbors
+from tests.common.helpers.constants import DEFAULT_ASIC_ID
 from tests.common.helpers.assertions import pytest_assert
 import re
 
 pytestmark = [
-    pytest.mark.topology('t1')
+    pytest.mark.topology('t1'),
+    pytest.mark.disable_loganalyzer
```
Comment (abdosi) [inline tests/bgp/test_traffic_shift.py:11]: why is this added ?

File: tests/bgp/conftest.py [hunk 9]
```diff
@@ -220,29 +220,41 @@ def _setup_interfaces_t1(mg_facts, peer_count):
...
                     conn["neighbor_intf"] = "eth%s" % mg_facts["minigraph_port_indices"][member_intf]
+                    conn["namespace"] = mg_facts["minigraph_portchannels"][intf]["namespace"]
                 else:
                     conn["neighbor_intf"] = "eth%s" % mg_facts["minigraph_port_indices"][intf]
                 connections.append(conn)
 
             for conn in connections:
                 # bind the ip to the interface and notify bgpcfgd
-                duthost.shell("config interface ip add %s %s" % (conn["local_intf"], conn["local_addr"]))
+                namespace = '-n {}'.format(conn["namespace"]) if conn["namespace"] else ''
```
Comment (abdosi) [inline tests/bgp/conftest.py:248]: if we are planning to use loopback ip for bgpmon why we need to configure local address ip address and also neighbor address on ptf ?  Is there any other use case other then bgpmon

File: tests/bgp/bgp_helpers.py [hunk 10]
```diff
@@ -118,3 +135,35 @@ def verify_all_routes_announce_to_bgpmon(duthost, ptfhost):
...
+    Remove the bgp neigbors for a particular BGP instance
+    """
+    namespace = duthost.get_namespace_from_asic_id(asic_index)
+
+    # Convert the json formatted result of sonic-cfggen into bgp_neighbors dict
+    bgp_neighbors = json.loads(duthost.command("sudo sonic-cfggen {} -d --var-json {}".format('-n ' + namespace if namespace else '', "BGP_NEIGHBOR"))["stdout"])
+    cmds = []
+    for neighbor in bgp_neighbors.keys():
+        cmds.append('sudo sonic-db-cli {} CONFIG_DB del "BGP_NEIGHBOR|{}"'.format('-n ' + namespace if namespace else '', neighbor))
+    duthost.shell_cmds(cmds=cmds)
```
Comment (abdosi) [inline tests/bgp/bgp_helpers.py:150]: we can remove loop and sonic-cfggen call.
this seems to work:
sonic-db-cli CONFIG_DB keys "BGP_NEI*" | xargs  sonic-db-cli CONFIG_DB del

File: tests/bgp/bgp_helpers.py [hunk 11]
```diff
@@ -118,3 +135,35 @@ def verify_all_routes_announce_to_bgpmon(duthost, ptfhost):
...
+    Restore the bgp neigbors for a particular BGP instance
+    """
+    namespace = duthost.get_namespace_from_asic_id(asic_index)
+
+    # Convert the bgp_neighbors dict into json format after adding the table name.
+    bgp_neigh_dict = {"BGP_NEIGHBOR":bgp_neighbors}
+    bgp_neigh_json = json.dumps(bgp_neigh_dict)
+    duthost.shell("sudo sonic-cfggen {} -a '{}' --write-to-db".format('-n ' + namespace if namespace else '', bgp_neigh_json))
+
+    # Restart BGP instance on that asic
```
Comment (abdosi) [inline tests/bgp/bgp_helpers.py:168]: can we do config_reload instead ?

Comment (judyjoseph) [reply tests/bgp/bgp_helpers.py:166]: I did this because the current approach is quick and less time consuming than config reload, and in the test we can do it very specific to the BGP neighbors in that asic index.

File: tests/bgp/bgp_helpers.py (hunk 10 above)
Comment (judyjoseph) [reply tests/bgp/bgp_helpers.py]: Updated.

File: tests/bgp/test_traffic_shift.py (hunk 8 above)
Comment (judyjoseph) [reply tests/bgp/test_traffic_shift.py]: removed, it was i think added as part of testing.

File: tests/bgp/test_traffic_shift.py (hunk 7 above)
Comment (judyjoseph) [reply tests/bgp/test_traffic_shift.py:26]: This API was added to give flexibility to check the scenario of no neighbors per ASIC. With this support in a multi-asic DUT we need not do the loop and check for all ASIC's - as the behavior can be verified with on asic alone, here I use the first ASIC in single/multi-asic DUT.

**Review by abdosi — APPROVED** [review]
